src/host/alloc_cache.jl

```diff
@@ -71,14 +76,13 @@ end
 
 function unsafe_free!(cache::AllocCache)
     Base.@lock cache.lock begin
-        for (_, pool) in cache.busy
-            isempty(pool) || error(
-                "Invalidating allocations cache that's currently in use. " *
-                    "Invalidating inside `@cached` is not allowed."
-            )
+        for pool in values(cache.busy)
+            isempty(pool) || error("Cannot invalidate a cache that's in active use")
```

Comment:
Maybe check that the pool is empty **and** that the cache is in use (reading ScopedValue). Otherwise, if an exception happens during `@cached` it will raise this in the cache finalizer.

Reply:
That sounds questionable. Shouldn't `@cached` instead have a `try`/`catch` (well, the scope-less variety) that wipes the cache in the `finally` block, making sure the cache is always in a known-good state outside of the `@cached` block?